key-manager/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A crate used to derive keys useful for the Oxide control plane
+
+use std::collections::BTreeMap;
+
+use async_trait::async_trait;
+use hkdf::Hkdf;
+use secrecy::{ExposeSecret, Secret};
+use sha3::Sha3_256;
+use sled_hardware::DiskIdentity;
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+/// Input Key Material
+///
+/// This should never be used directly, and always wrapped in a `Secret<Ikm>`
+/// upon construction. We sparate the two types, because a `Secret` must contain
+/// `Zeroizable` data, and a `Box<[u8; 32]>` is not zeroizable on its own.
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct Ikm(pub Box<[u8; 32]>);
```

Comment:
I assume the omission of `Debug` impls is intentional. Would it be useful to implement `Debug` manually in a way that does not reveal any data? I can think of a couple benefits but am not sure it's completely worth it:

1. types containing any of these values can derive `Debug`
2. it's very obvious that the `Debug` impl is special to avoid leaking information

Reply:
It's definitely intentional. I'm also not sure why you'd want a `Debug` impl here. This is a secret and we don't want to log it by accident. The secret is further protected in the `Secret` type below.

Reply:
So one other thing here. The secrets themselves are meant to be transient, and shouldn't be stored in any longer lived data structure. The IKM gets created once and put in a secret, where it is fed into Hkdf and discarded.

Reply:
It's a good point. I'll just derive it to avoid the comment and problems :)

Reply:
> It's a good point. I'll just derive it to avoid the comment and problems :)

I'll **impl** it, not derive. Sigh.

Reply:
Fixed in 6abfb00

Reply:
I actually removed the `Debug` impl again and instead made this type private in [79cf577](https://github.com/oxidecomputer/omicron/pull/2990/commits/79cf577eda5264af8622c8e48299c706b2a89a8e)